contrib/udfs/src/main/java/org/apache/drill/exec/udfs/UserAgentAnalyzerProvider.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.drill.exec.udfs;
+
+import nl.basjes.parse.useragent.UserAgentAnalyzer;
+
+public class UserAgentAnalyzerProvider {
+
+  public static UserAgentAnalyzer getInstance() {
+    return UserAgentAnalyzerHolder.INSTANCE;
+  }
+
+  private static class UserAgentAnalyzerHolder {
+    private static final UserAgentAnalyzer INSTANCE = UserAgentAnalyzer.newBuilder()
```

Comment:
Jumping in a bit late here. Isn't one layer of class sufficient to hold the static instance?

```
public class UserAgentAnalyzerProvider {
  private static final INSTANCE = UserAgentAnalyzer.newBuilder()
	    .dropTests()
            .hideMatcherLoadStats()
            .immediateInitialization()
            .build();

  public static UserAgentAnalyzer getInstance() {
    return INSTANCE;
  }
}
```

Reply:
@paul-rogers may I answer this question.
Your suggestion will also work but will contradict with holder on demand singleton approach, i.e. it's not lazy init approach. Your example will init singleton instance when JVM starts while having inner class will init singleton only during first call of `getInstance` method.

Reply:
@arina-ielchiieva, thanks for the explanation. I understand the reasoning. However, what I see, at least in the debugger, is that statics are not initialized until the first use of the class. A reference is not a use, only a call is.

However, this behavior may be unique to the debugger, perhaps the non-debug execution works differently. Still, easy enough to test so we can know for certain. Would be good to nail down the exact behavior so we don't add excess complexity unnecessarily.

Reply:
Did a little experiment to see if I was confused. Looks like a single layer of provider will work fine.

Two classes:

```
public class DummyProvider {
  private static int instance = 10;
  static {
    System.out.println("DummyProvider loaded");
  }

  public static int instance() { return instance; }
}
```

And:

```
public class DummyClient {
  public static void main(String[] args) {
    System.out.println("Main starting");
    System.out.println("Getting instance");
    System.out.println(DummyProvider.instance());
    System.out.println("Exiting");
  }
}
```

Output:

```
Main starting
Getting instance
DummyProvider loaded
10
Exiting
```

Ran in both the debugger and normally. Both produced the same output.

This suggests that, indeed, the class is not loaded, and statics initialized, until first use (call).

Reply:
@paul-rogers here is some explanation that you might find useful: https://en.m.wikipedia.org/wiki/Initialization-on-demand_holder_idiom

Reply:
@arina-ielchiieva, thanks; that article is accurate in the case that class `Something` is loaded earlier than the first use of `getInstance()`. Maybe there are other static functions or constants that force earlier class loading.

In this specific case, the only method is `getInstance()` and so the outer class won't be loaded until that time. This means that the inner and outer classes are load at essentially the same time: on that first call to `getInstance()`.

We can commit his PR with the code as it is; it works fine. My concern, however, is that we use static instances all over, and we've kind of relied on the behavior I've been describing. To be consistent, we'd want to go and change all the other uses as well. But, doing so would be unnecessary work in cases, like this one, where a single layer works fine.

Let's go ahead and commit this, then we can continue the discussion without slowing this PR.

Reply:
@paul-rogers totally agree with you, since there is not other static stuff there class won't be loaded prematurely, unless with time something static is added there. Anyway, having inner class does not give much complexity and since you ok, we can leave it as it to comply with pattern code template.